### PR TITLE
fix(resolver/native/incremental): callerName parity, func-prop cross-file edges, O(n) Phase 8.3f, qualified callerName dispatch

### LINKED ISSUE #1370: fix(incremental): pass callerName to resolveCallTargets for Phase 8.3f scoped typeMap lookup
## Problem

In `incremental.ts:513-519`, `resolveCallTargets` is called without the `callerName` parameter:

```ts
const { targets, importedFrom } = resolveCallTargets(
  lookup,
  call,
  relPath,
  importedNames,
  typeMap,
  // callerName NOT passed
);
```

`resolveByMethodOrGlobal` has a third lookup branch (added in #1368, Phase 8.3f) that tries the scoped key `${callerName}::${effectiveReceiver}`. Without `callerName`, this branch is never reached during incremental watch-mode rebuilds.

## Impact

When two functions in the same file share the same rest-param name (e.g. `...rest`), Phase 8.3f seeds scoped typeMap keys (`f1::rest`, `f2::rest`) and only seeds the unscoped key when exactly one callee uses that rest name. During an incremental rebuild, the scoped lookup is skipped — so in the two-callee collision scenario, neither function resolves its rest-param calls via the incremental path.

A full rebuild always produces the correct graph; the regression is limited to watch-mode incremental updates on the affected files.

## Fix

Pass `caller.callerName` as the 6th argument to `resolveCallTargets` at `incremental.ts:513`:

```ts
const { targets, importedFrom } = resolveCallTargets(
  lookup,
  call,
  relPath,
  importedNames,
  typeMap,
  caller.callerName,  // add this
);
```

Note: `findCaller` is called on line 512, so `caller.callerName` is already available at that point.

## Context

Identified during review of PR #1368. Pre-existing gap in incremental Phase 8.3f coverage, widened by the scoped-key change.

### LINKED ISSUE #1371: fix(native): func-prop post-pass misses call edges in files that only call but don't define func-prop methods
## Problem

The `runPostNativePrototypeMethods` post-pass in `native-orchestrator.ts` pre-filters files to only those that match the func-prop regex (`fn.method = function(){}` or arrow form). This means the WASM re-parse only covers definition files.

A file that only *calls* `f.method()` without also defining any `\w+.\w+ = function` pattern is absent from `wasmResults` and its call edges to newly-inserted func-prop nodes are never emitted.

**Concrete example:**
- `lib.js` defines `f.method = function(){}` → f.method node is inserted
- `app.js` only calls `f.method()` → this edge is silently dropped

The comment at the top of the loop says "Resolve call edges in every file — not just those that define new func-prop methods", which contradicts the implementation.

## Fix options

1. Extend the WASM parse to include ALL JS/TS files (higher cost but correct)
2. After inserting new nodes, query the DB for raw call sites that match the receiver name/method name pattern of newly-inserted nodes (DB-based, no re-parse needed)

## Impact

Low-frequency in practice: most codebases that use `fn.method = function(){}` tend to define and call in the same file or module. But multi-file patterns are a legitimate use case.

## Context

Identified in Greptile review of PR #1331, comment https://github.com/optave/ops-codegraph-tool/pull/1331#discussion_r3367103832

### LINKED ISSUE #1374: perf(resolver): index paramBindings and objectPropBindings to reduce Phase 8.3f from O(n³) to O(n)
## Problem

The Phase 8.3f resolution in `buildPointsToMap` (`src/domain/graph/resolver/points-to.ts:187-199`) uses three nested `for...of` loops with no early indexing:

```ts
for (const restBinding of objectRestParamBindings) {
  for (const paramBinding of paramBindings) {
    for (const propBinding of objectPropBindings) {
      if (...equality guards...) { ... }
    }
  }
}
```

In a large generated or bundled file with many rest-param functions and object literals, the inner-loop body executes on every combination before the equality guards discard most pairs. This is O(|objectRestParamBindings| × |paramBindings| × |objectPropBindings|).

## Fix

Pre-index the bindings before the loops:
- `paramBindings` as `Map<callee+argIndex, argName[]>`
- `objectPropBindings` as `Map<objectName, {propName, valueName}[]>`

This reduces the lookup to O(n) and avoids the quadratic scan of `paramBindings` for each rest binding.

## Impact

Low risk for current benchmark files (small fixtures). Becomes a performance concern for large real-world codebases with many rest-destructured parameters.

## Context

Identified in Greptile review of PR #1331.

### LINKED ISSUE #1385: fix(resolver): qualified callerName mismatch in class-scoped typeMap lookup
## Background

`resolveByMethodOrGlobal` in `src/domain/graph/builder/call-resolver.ts` uses
`callerName.lastIndexOf('.')` to extract the class prefix when doing class-scoped
`this.prop` typeMap lookups (introduced in #1382).

## Problem

The extractor (`handlePropWriteTypeMap`) stores keys using the bare class name from
`nameNode.text` — e.g. `ClassA.prop`. The resolver uses `lastIndexOf('.')` which
returns the full prefix for callerNames with more than one dot segment:

- `callerName = 'Namespace.ClassA.method'` → `callerClass = 'Namespace.ClassA'`
- Lookup: `typeMap.get('Namespace.ClassA.prop')` — will never match `ClassA.prop`

The same `lastIndexOf` pattern is already present for same-class method dispatch at
line 184, so this is a pre-existing limitation of the naming scheme rather than a
new regression introduced by #1382.

## Impact

Only affects codebases that produce namespace-qualified callerNames in the graph
(e.g. `Namespace.ClassName.method`). In practice most JS/TS graphs emit bare
`ClassName.method` format, so real-world impact is currently minimal.

## Fix

When `callerName` contains multiple dots, extract only the last dotted segment as
the class name (i.e. use the segment immediately before the method name), consistent
with how the extractor names the key. This applies to both the this-prop lookup
(line 82) and the same-class method dispatch (line 184).

## Related

- #1323 (original fix), PR #1382
- Same `lastIndexOf` pattern at line 184 (`resolveByMethodOrGlobal`)

### PR DESCRIPTION
**Pull request by carlos-alm**
## Summary

Five fixes in the native/resolver path, all identified in Greptile reviews:

**Fix 1 (#1370): pass callerName in incremental path**
- `incremental.ts:513` now passes `caller.callerName` to `resolveCallTargets` (was previously omitted).
- All callerName-dependent resolution paths in `resolveByMethodOrGlobal` were silently skipped during watch-mode incremental rebuilds: class-scoped `this.prop`, same-class `this.method()`, Object.defineProperty accessor `this`-dispatch.

**Fix 2 (#1371): second WASM pass for caller-only files in func-prop post-pass**
- `runPostNativePrototypeMethods` only WASM-parsed definition files, so call edges from files that only *called* `f.method()` were silently dropped.
- Adds a second WASM pass after node insertion (text-searches remaining JS/TS files for matching method-name patterns) and a direct `receiver.method` qualified-name lookup fallback for untyped receivers.
- Escapes regex special characters in method name suffixes so names like `$init` or `set+value` don't break the text pre-filter (#1383 follow-up).

**Fix 3 (#1374): O(n³) → O(n) for Phase 8.3f in `buildPointsToMap`**
- The object-rest parameter dispatch block had three nested `for...of` loops with equality guards that discarded most pairs only after iterating all combinations.
- Pre-indexes `paramBindings` as `Map<"callee::argIndex", argName[]>` and `objectPropBindings` as `Map<objectName, props[]>` — both built in O(n), reducing the full block to O(n) lookups.

**Fix 4 (#1385): extract bare class name from qualified callerName in same-class dispatch**
- `resolveByMethodOrGlobal` used `callerName.lastIndexOf('.')` to derive the class prefix for same-class `this.method()` dispatch.
- For namespace-qualified callerNames like `Namespace.ClassName.method`, this yielded `Namespace.ClassName` — but symbols are stored under the bare name `ClassName`, so the lookup silently returned empty.
- Fix: walk one dot further back (`lastIndexOf('.', dotIdx - 1)`) to extract only the immediate class segment.

## Test plan
- [ ] `tests/integration/incremental-callerName-parity.test.ts` — callerName incremental parity (#1370)
- [ ] `tests/integration/issue-1370-incremental-caller-name.test.ts` — callerName watch-mode regression (#1370)
- [ ] `tests/integration/func-prop-caller-only-file.test.ts` — cross-file edge from caller-only file (#1371)
- [ ] `tests/integration/func-prop-this-dispatch.test.ts` — existing same-file func-prop test (regression guard)
- [ ] `tests/unit/points-to.test.ts` — 7 new Phase 8.3f unit tests + 16 existing (#1374)
- [ ] `tests/unit/call-resolver.test.ts` — 4 new unit tests for qualified callerName dispatch (#1385)

Closes #1370
Closes #1371
Closes #1374
Closes #1385

### CODE REVIEW

**Comment by carlos-alm** [comment]
@greptileai

**Comment by carlos-alm** [comment]
Tracked the remaining incremental/full-build divergence (same-class `this.method()` fallback + Object.defineProperty accessor fallback) in #1384 as recommended.

File: src/domain/graph/builder/stages/native-orchestrator.ts [hunk 1]
```diff
@@ -669,6 +669,47 @@ async function runPostNativePrototypeMethods(
...
+  if (newMethodSuffixes.size > 0) {
+    const protoFileSet = new Set(protoFiles);
+    const callerCandidateAbs: string[] = [];
+    for (const relPath of jsFiles) {
+      if (protoFileSet.has(relPath)) continue; // already parsed in first pass
+      try {
+        const content = readFileSafe(path.join(rootDir, relPath));
+        const matchesAny = [...newMethodSuffixes].some((m) =>
+          new RegExp(`\\.${m}\\s*\\(`).test(content),
+        );
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply src/domain/graph/builder/stages/native-orchestrator.ts]: Fixed in commit 0eb7a5a. The method suffix is now escaped before being interpolated into `new RegExp()`, so method names containing `$` or other regex metacharacters are handled correctly: `const escaped = m.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')`.

**Comment by carlos-alm** [comment]
@greptileai

**Comment by carlos-alm** [comment]
Addressed the two items from the Greptile review summary (commit 69bd0bc):

- **Redundant `call.receiver &&` guard** (lines 789-795): removed — the outer `if (!call.receiver) continue` at line 774 already ensures receiver is always truthy at that point. Added a comment explaining the invariant.
- **RegExp re-compilation inside `.some()`** (lines 688-697): pre-compile all suffix patterns into `suffixPatterns` once before the file scan loop, then call `suffixPatterns.some((re) => re.test(content))` per file — eliminates up to `newMethodSuffixes.size` regex compilations per file.

**Comment by carlos-alm** [comment]
@greptileai

**Comment by carlos-alm** [comment]
Pushed 3 additional commits that address issue #1388: the `more1` hand-authored fixture (array iteration patterns) has been moved from `jelly-micro/` to a new `pts-javascript/` fixture directory. The fixture is now split into 4 focused single-pattern files (for-of.js, set-iter.js, array-from.js, spread.js) to eliminate intra-file cross-pollination. All 13 expected edges pass at 100% precision/100% recall.

The Windows CI failure on the previous CI run was a transient HTTP 504 from GitHub's CDN for tree-sitter-windows-x64.gz (not caused by this PR). New CI run triggered by this push.

Also filed #1418 for a pre-existing C# build-parity test failure (native vs WASM role/edge divergence on main) that is unrelated to this PR.

**Comment by carlos-alm** [comment]
@greptileai
